v4/export/sql.go

```diff
@@ -1009,3 +1056,150 @@ func GetPartitionNames(db *sql.Conn, schema, table string) (partitions []string,
+			dbName:        dbName,
+			tbName:        tbName,
+			partitionName: partitionName,
+		}
+	}
+	err = simpleQuery(db, selectStatsHistogramsSQL, func(rows *sql.Rows) error {
+		var (
+			tableID                   int64
+			updateTime, distinctCount string
+		)
+		err2 := rows.Scan(&tableID, &updateTime, &distinctCount)
+		if err2 != nil {
+			return errors.Trace(err2)
+		}
+		if mpt, ok := saveMap[updateTime]; ok {
+			if partition, ok := mpt[distinctCount]; ok {
+				dbName, tbName, partitionName := partition.dbName, partition.tbName, partition.partitionName
+				if _, ok := partitionIDs[dbName]; !ok {
+					partitionIDs[dbName] = make(map[string]map[string]int64)
+				}
+				if _, ok := partitionIDs[dbName][tbName]; !ok {
+					partitionIDs[dbName][tbName] = make(map[string]int64)
+				}
+				partitionIDs[dbName][tbName][partitionName] = tableID
+			}
+		}
+		return nil
+	})
+	return partitionIDs, err
+}
+
+// GetDBInfo get model.DBInfos from database sql interface.
+// We need table_id to check whether a region belongs to this table
+func GetDBInfo(db *sql.Conn, tables map[string]map[string]struct{}) ([]*model.DBInfo, error) {
+	const tableIDSQL = "SELECT TABLE_SCHEMA,TABLE_NAME,TIDB_TABLE_ID FROM INFORMATION_SCHEMA.TABLES ORDER BY TABLE_SCHEMA"
+
+	schemas := make([]*model.DBInfo, 0, len(tables))
+	var (
+		tableSchema, tableName string
+		tidbTableID            int64
+	)
+	partitionIDs, err := GetPartitionTableIDs(db, tables)
+	if err != nil {
+		return nil, err
+	}
+	err = simpleQuery(db, tableIDSQL, func(rows *sql.Rows) error {
+		err2 := rows.Scan(&tableSchema, &tableName, &tidbTableID)
+		if err2 != nil {
+			return errors.Trace(err2)
+		}
+		if tbm, ok := tables[tableSchema]; !ok {
+			return nil
+		} else if _, ok = tbm[tableName]; !ok {
+			return nil
+		}
+		last := len(schemas) - 1
+		if last < 0 || schemas[last].Name.O != tableSchema {
+			schemas = append(schemas, &model.DBInfo{
+				Name:   model.CIStr{O: tableSchema},
+				Tables: make([]*model.TableInfo, 0, len(tables[tableSchema])),
+			})
+			last++
+		}
+		var partition *model.PartitionInfo
+		if tbm, ok := partitionIDs[tableSchema]; ok {
+			if ptm, ok := tbm[tableName]; ok {
+				partition = &model.PartitionInfo{Definitions: make([]model.PartitionDefinition, 0, len(ptm))}
+				for partitionName, partitionID := range ptm {
+					partition.Definitions = append(partition.Definitions, model.PartitionDefinition{
+						ID:   partitionID,
+						Name: model.CIStr{O: partitionName},
+					})
+				}
+			}
+		}
+		schemas[last].Tables = append(schemas[last].Tables, &model.TableInfo{
+			ID:        tidbTableID,
+			Name:      model.CIStr{O: tableName},
+			Partition: partition,
+		})
+		return nil
+	})
+	return schemas, err
+}
+
+// GetRegionInfos get region info including regionID, start key, end key from database sql interface.
+// start key, end key includes information to help split table
+func GetRegionInfos(db *sql.Conn) (*helper.RegionsInfo, error) {
```

Comment:
not sure querying all info will cost too much memory of TiDB and dumpling

Reply:
These are just metadata, won't cost too much memory.
Actually in v4.0 TiDB will also grab data from PD at one time and then filter these data.